DOCSP-33631: multi-field join tutorial (#834)
mongodb/docs-node: source/aggregation-tutorials/multi-field-join.txt

Multi-Field Join
================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

.. facet::
   :name: genre
   :values: tutorial

.. meta::
   :keywords: code example, node.js, lookup, aggregation

Introduction
------------

In this tutorial, you can learn how to use the {+driver-short+} to
perform a multi-field join. A multi-field join occurs when there are

> multiple fields in the documents of two collections that you use to

Reviewer: S: I think adding "corresponding" clarifies why/how you can use these multiple fields to match documents together

```suggestion
multiple corresponding fields in the documents of two collections that you use to
```

field values and combines information from both into one document.

.. tip:: One-to-many Joins

   A one-to-many join is a variety of a multi-field join. When you
   perform a one-to-many join, you select one field from a document that
   matches a field value in multiple documents on the other side of the
   join. To learn more about these data relationships,
   see the Wikipedia entries about :wikipedia:`One-to-many (data model)
   <w/index.php?title=One-to-many_(data_model)&oldid=1112674599>` and
   :wikipedia:`Many-to-many (data model)
   <w/index.php?title=Many-to-many_(data_model)&oldid=1169943560>`.

Aggregation Task Summary
~~~~~~~~~~~~~~~~~~~~~~~~

This aggregation tutorial demonstrates how to join data from two
collections related to product information and customer orders.
The aggregated result is a list of products for which orders were made
in 2020.

This example uses two collections:

- ``products``, which contains documents describing the products that
  a shop sells
- ``orders``, which contains documents describing individual orders
  for products in a shop

An order can only contain one product, so the aggregation uses a

> multi-field join to match a product document to documents for orders of

Reviewer: S: I think "for" is a little unclear
```suggestion
multi-field join to match a product document to documents representing orders of
```

               :language: javascript
               :copyable: true
               :start-after: start-embedded-pl-match1
               :end-before: end-embedded-pl-match1
               :dedent:

            Within the embedded pipeline, add another :manual:`$match
            </reference/operator/aggregation/match>` stage to match
            orders placed in 2020:

            .. literalinclude:: /includes/aggregation/multi-field-join.js
               :language: javascript
               :copyable: true
               :start-after: start-embedded-pl-match2
               :end-before: end-embedded-pl-match2
               :dedent:

            Within the embedded pipeline, add an :manual:`$unset
            </reference/operator/aggregation/unset>` stage to remove
            unneeded fields from the ``orders`` collection side of the join:

            .. literalinclude:: /includes/aggregation/multi-field-join.js
               :language: javascript
               :copyable: true
               :start-after: start-embedded-pl-unset
               :end-before: end-embedded-pl-unset
               :dedent:

            After the embedded pipeline is completed, add the

>             ``$lookup`` stage to the main aggregation pipeline that also stores
>             the processed lookup fields in the ``orders`` array field:

Reviewer: S: breaking up the sentence might help make it more action-oriented

```suggestion
            ``$lookup`` stage to the main aggregation pipeline. Configure this stage to store
            the processed lookup fields in the ``orders`` array field:
```

               :copyable: true
               :start-after: start-lookup
               :end-before: end-lookup
               :dedent:

         .. step::

            Next, add a :manual:`$match
            </reference/operator/aggregation/match>` stage to only show
            products for which there is at least one order in 2020,
            based on the ``orders`` array calculated in the previous step:

            .. literalinclude:: /includes/aggregation/multi-field-join.js
               :language: javascript
               :copyable: true
               :start-after: start-match
               :end-before: end-match
               :dedent:

         .. step::

            Finally, add an :manual:`$unset
            </reference/operator/aggregation/unset>` stage. The
            ``$unset`` stage removes the ``_id`` field from the result documents:
            
            .. literalinclude:: /includes/aggregation/multi-field-join.js
               :language: javascript
               :copyable: true
               :start-after: start-unset
               :end-before: end-unset
               :dedent:

   .. step:: Run the Aggregation Pipeline

>       Perform the aggregation on the ``products`` collection by running the
>       following command in your shell to run the ``agg_tutorial.js``
>       file:

Reviewer: S: Not necessary, but I found the current wording slightly confusing
```suggestion
      Run the `agg_tutorial.js`` file to perform the aggregation on the ``products`` collection by running the
      following command in your shell:
```

Author: Simplified and changed in both existing tutorials

      
         node agg_tutorial.js

      The aggregated result contains two documents. The documents
      represent products for which there were orders placed in 2020.
      Each document contains an ``orders`` array field that lists details
      about each order for that product:

      .. code-block:: javascript
         :copyable: false
         
         {
           name: 'Asus Laptop',
           variation: 'Standard Display',
           category: 'ELECTRONICS',
           orders: [
             {
               customer_id: 'elise_smith@myemail.com',
               orderdate: 2020-05-30T08:35:52.000Z,
               value: 431.43
             },
             {
               customer_id: 'jjones@tepidmail.com',
               orderdate: 2020-12-26T08:55:46.000Z,
               value: 429.65
             }
           ]
         }
         {
           name: 'Morphy Richards Food Mixer',
           variation: 'Deluxe',
           category: 'KITCHENWARE',
           orders: [
             {
               customer_id: 'oranieri@warmmail.com',
               orderdate: 2020-01-01T08:25:37.000Z,
               value: 63.13
             }
           ]
         }

>       The result consists of documents that contain details from

Reviewer: S: just to shorten
```suggestion
      The result documents contain details from
```
